### DISCUSSION TITLE
20Th hashpower can't successfully get even 1 block of Ethereum

### DISCUSSION THREAD

**Discussion by James1847** [coinfoundry/miningcore#1146] [Q&A]
We use nicehash for our miner, and we send 20TH hashpower to mine the block of Ethereum which should get a block every 5 min, but in reality, no block mined from our mining pool, what's  going on? Here is the configuration file:

_{
  "logging": {
    "level": "info",
    "enableConsoleLog": true,
    "enableConsoleColors": true,
    "logFile": "",
    "logBaseDirectory": "",
    "perPoolLogFile": false
  },
  "banning": {
    "manager": "Integrated",
    "banOnJunkReceive": true,
    "banOnInvalidShares": false
  },
  "notifications": {
    "enabled": false,
    "email": {
      "host": "smtp.example.com",
      "port": 587,
      "user": "user",
      "password": "password",
      "fromAddress": "info@yourpool.org",
      "fromName": "support"
    },
    "admin": {
      "enabled": false,
      "emailAddress": "user@example.com",
      "notifyBlockFound": true
    }
  },
  "persistence": {
    "postgres": {
      "host": "127.0.0.1",
      "port": 5432,
      "user": "miningcore",
      "password": "some-secure-password",
      "database": "miningcore"
    }
  },
  "paymentProcessing": {
    "enabled": true,
    "interval": 100,
    "shareRecoveryFile": "recovered-shares.txt"
  },
  "api": {
    "enabled": true,
    "listenAddress": "0.0.0.0",
    "port": 4000,
    "metricsIpWhitelist": [
      "18.162.57.76"
    ],
    "rateLimiting": {
      "disabled": true
    }
  },
  "pools": [
    {
      "id": "eth1",
      "enabled": true,
      "coin": "ethereum",
      "address": "0x6486A3f7849734d0EA333eaf9244A4A34dCA17f6",
      "rewardRecipients": [
        {
          "type": "op",
          "address": "0x6486A3f7849734d0EA333eaf9244A4A34dCA17f6",
          "percentage": 0
        }
      ],
      "enableDaemonWebsocketStreaming": false,
      "blockRefreshInterval": 50,
      "clientConnectionTimeout": 600,
      "banning": {
        "enabled": true,
        "time": 600,
        "invalidPercent": 50,
        "checkThreshold": 50
      },
      "ports": {
        "6161": {
          "listenAddress": "*",
          "difficulty": 3,
          "varDiff": null
        }
      },
      "daemons": [
        {
          "host": "127.0.0.1",
          "port": 8545,
          "portWs": 8546,
          "notifyWorkUrl": "http://127.0.0.1:8107"
        }
      ],
      "paymentProcessing": {
        "enabled": true,
        "minimumPayment": 0.01,
        "payoutScheme": "SOLO",
        "payoutSchemeConfig": {
          "factor": 0.5
        },
        "gas": 21000,
        "maxFeePerGas": 50000000000,
        "keepUncles": false,
        "keepTransactionFees": false
      }
    }
  ]
}
_

**Comment by XIII39** [comment]
It's the luck of the math. How many minutes/hours are you renting 20TH for from NH?

Currently in order to solve a block the GPU needs to find a solution equivalent to D=2,882,421.42 or greater, 12.38 P at the time I'm posting this

Have a look at this wallet - https://solo-eth.2miners.com/account/0x0e3918efec28549af51a80f7776d0a75783083ec#rewards. From what I can see, this person often sends 15TH - 20TH though out the day and if you look at the "personal luck" there are times when it is well over 200% ever 400%. Even though it takes longer than expected to find a block, the one thing to keep in mind is that over the long-term, say 365 days, you will average a 100% luck/probability factor. 

In my personal experience, my network will constantly solve 3 block (last on November 1st, 10th & 11th) but then will not solve another block for 60-80 days, then it does it again.. 2 or 3 blocks.

I track the actual difficulty of each share that my pool receives so I can see the math. I can tell you that the last time I rented 6TH from NH from 1h 20m the largest share received was 5.04P. 6TH should, statically, solve 1.5 blocks per hour but, I just didn't have the luck that day.

**Reply by PromediaB** [reply]
Thank you for your answer, its making clear a part of question, but before waiting the luck, we should be sure that our configuration is fine. as we read soo much article that talking about the configuration.

**Reply by XIII39** [reply]
Until a GPU finds an acceptable solution with a difficulty high enough to be considered a block solution you wont really know if there is a configuration problem.

Your Miningcore log file will show something like this for a successful solution
```
Submitting block <BLOCK_NUMBER> 
Daemon accepted block <BLOCK_NUMBER> submitted by <WALLET_ID>
```
or for an unsuccessful solution - I'm not 100% sure what other detail might be here, I've never had a rejected block myself. Making a bit of an assumption here. 
```
Submitting block <BLOCK_NUMBER> 
Daemon rejected block <BLOCK_NUMBER> submitted by <WALLET_ID>
```

An unsuccessful solution could be caused by a number of factors like your node wasn't fast enough to populate the block so someone beat you to it, the difficulty was close but not high enough, your network was slow and there were already a maximum number of Uncle block, etc...

If you want to gain some confidence in your configuration, I would suggest that you run your pool on the Ropsten testnet, solve a few blocks, check your logs so you know what you're looking for and check your "mined blocks" and "mined uncles" on https://ropsten.etherscan.io/ to varify that everything is working before running on the Mainnet.

**Reply by PromediaB** [reply]
Thank you for your wallet explain, we will check log and understand the number. 

One question about the difficulty in the config file. Should we use default config or 4 or maybe increase it to 16 or 32 ?

what will thé différence ?

**Reply by XIII39** [reply]
The difficulty setting is just for the pool to keep a record of the shares. The only shares that matter are ones that are high enough to be a solution.

The higher the difficulty, the less often the GPUs will 'report' to the pool. 
Difficulty is calculated at a factor of 2^32 or 4.295 GH so
D=1 is 4.295 GH
D=4 is 17.18 GH
D=16 is 68.72 GH and so on.

Let's say you set the difficulty to 16, all this means is that the mining GPU will report to the pool whenever it finds a share that is 68.72 GH or greater. The pool difficulty has nothing to do with finding a block solution. The higher the difficulty the less network stress on your pool/network.

If you set your difficulty to 2,882,421.42 (which some mining software like Trex really don't like) the GPU would only report to the pool when a share is 12.38 PH which, today, would be a block solution. Mind you if the network difficulty drops to say 11 PH you and you find a solution that is 11 PH but the difficulty is set to 12+, Minincore wouldn't know and the block would go unsolved.

I use D=16 for my Nicehash connections because it provides enough shares per minute to keep my OCD brain at ease. Anything less and the submitted shares per second is just too many to look at, also it tends to cause Segment Faults on the old unix box I'm using.. Too much for the memory I guess.

**Reply by XIII39** [reply]
I guess I should add that the difficulty at the pool is primely used in order to calculate the percentage that a given worker has contributed to a block solution so they can be paid accordingly.. 

Let's say the difficulty is set to 4 so each share is 17.18 GH or better. Now the pool considers each share to be 17.18 GH
A block is found and the difficulty to find that block was 12 PH and the reward is 2 ETH 
At 12 PH and a difficulty of 4 it should take 698,486.61 shares to find the block (12PH / 17.18GH), that would be 100% luck.

3 miners are working on the pool.
Miner 1 sent 2,000 shares
Miner 2 sent 10,000 shares
Miner 3 sent 25,000 shares

We can take the reward of 2 ETH and divide it by the number of shares required = 2 ETH / 698,486.61 shares = 2.8633333429255e-6 ETH per share.

Miner 1 earns 0.005726 ETH
Miner 2 earns 0.028633 ETH
Miner 3 earns 0.071583 ETH

As you can see the pool won a block earlier than expected, the luck here is about 5% (37,000 shares out of 698,486.61)

The remaining ETH from the block reward should stay in the pool wallet and be saved by the pool operator to pay miners when the pool luck is over 100% - keeps your participants happy ;)

**Reply by PromediaB** [reply]
Thank you for your completed answer, could you check our graph, even after disabling vardiff by using D=16 in password, we continue to have some rejection.

We just got our block after the change.

<img width="1097" alt="Capture d’écran 2022-01-30 à 23 05 25" src="https://user-images.githubusercontent.com/51176677/151719856-43eea27b-7a82-4ada-8e97-bbb3a71da601.png">

I want also to ask you if have to make some pause of Hash after getting block, for exemple stop the big hash for 30 min after getting block. or this cannot impact anything for the new coming bloc ?
.

**Reply by XIII39** [reply]
The red you are seeing is not rejected at the pool but the delta at the pool.
If you had rejected there would be a red line on the upper half of the graph, where it is showing the Price, Spent, Speed Limit, etc...

The delta at the pool is when you receive more (green) or less (red) hash rate then you are paying for.

You buy 1 TH, the pool reports to receive 1.01 TH - Green, positive delta of 0.01 TH
You buy 1 TH, the pool reports to receive 990 GH - Red, negative delta of 10 GH

I don't see the need to pause after solving a block. Remember only 1 correct solution is required to solve a block. If there are 200 GPUs attempting to solve a block, their attempts don't change because one of them found a solution. In fact, they are so busy attempting to solve that they have no idea that someone in their family just solved one.. They just get new work and keep going.

I have solved multiple blocks within minutes by not only the same rig but the same GPU.. That 3080 was having a good day, Ha!

**Reply by PromediaB** [reply]
All this information was really helpful and removed a lot of cloud from our vision. we willl surrely get more experience by time.

We are trying to achieve the same goal as : https://miningdao.io/blocks  but even running the same hash we are too faraway from their resullt. 

Do you think that's because of configuration ? or do we need a time to warmup our pool for exemple ?

**Reply by XIII39** [reply]
It's all about time. Some days you will solve 0 block, others 5 or 6.. Sometimes even 10
As I mentioned before, in the long-term you will achieve a 100% luck factor, i.e. average X number of ETH/day based on your hash rate but that comes with time and patience. 

The one thing I really love about crypto is the math.. It never lies. Men lie, Women lie but math never lies. Ha!
I could get deep into calculating pool hashrate vs. network hashrate vs network difficulty to find the percentage probability to solve a block based on realtime + blocktime but the basic calculators online give enough information to ballpark your time to solve.

Again, my OCD.. I like to know the minute my chance of solving a block goes from 0% to 0.01%. It's how I ignore the rest of the world around me.. Some people use TikTok, I use math and mining. Haha..

Now, on to the next block! :)

**Reply by PromediaB** [reply]
That's mean the time of getting a block can be predicted with some math ? that can help us to put off the hash 80% and run it 20% if we get the good calculation ?

**Reply by XIII39** [reply]
It is possible but not as much when you're renting hashrate, I've tried to predict it many times.
One of the constants we need is the size of the mining network. Rented hashrate changes that variable. Also because we are working with time, both real time and block time, it makes it very difficult to do calculations with 1 hour of 1TH or 30min of 5TH, you know what I mean?

**Comment by konstiantyn** [comment]
Not a block was found at a hashrate of 40GH/s in the last 20 days.

**Reply by XIII39** [reply]
Statistically you should have won between 4-5 blocks but until day 22 there is a 1% chance you will have 0 blocks.
My guess is that you will solve 4 or 5 blocks in close succession in the next few days..

As an example lets say you have 40GH of hashrate and we know that right now the network hashrate is 1,004,974.89 GH and the block time is 14s (rounded for ease). Since we know the number of events and the frequency of the events, I would use Poisson Distribution (https://en.wikipedia.org/wiki/Poisson_distribution) to calculate probability.

Convert everything to MH/s - I'm giving these variables a,b,c,d which I will use in the math formulas below
a) 40GH = 40,000 MH (miner)
b) 1,004,974.89 GH = 1,004,974,890 MH (network hashrate)
c) Block Time = 14 seconds
d) Mining Time = 1,728,000 seconds (20 days)

For Poisson Distribution first we need to find Lambda.
(a/b)*(d/c) = Lambda

Then we use Lambda as the mean value in the Poisson Distribution formula and find the percentage chance of a solve blocks..
So 40GH mining for 20 full days:
0 Blocks = 1%
1 Block = 4%
2 Blocks = 9%
3 Blocks = 15%
4 Blocks = 18%
5 Blocks = 18%
6 Blocks = 14%
7 Blocks = 10%
8 Blocks = 6%
9 Blocks = 3%
10 Blocks = 2%
11 Blocks = 1%
12 Blocks = 0%
 
Here was can see that if we mine with 40GH for 20 days the probability is that we will solve between 4-5 blocks. If we are extremely lucky the we could solve up to 11 and if we are extremely unlucky then we can solve as little as 0. On day 22 solving 0 blocks becomes 0%

I would say that if you don't solve a block over the next few days, then there could be is something wrong with our network/node. Ping time, speed, etc... but chances are that you will solve many blocks very quickly in a cluster.

You can also use Binomial Distribution, calculate the probability per block against time and yield the same results
Probability per block would be calculated by
a/b=probability

**Reply by alexanderkiehl** [reply]
Hello,

Could we assume that if we let 100 Gh run all the time and run multiple time rent power for example 15th for 15 min each 2 hours could we define the best time to launch and optimize our luck to catch somes blocks ?

**Reply by XIII39** [reply]
Not necessary. There is no way to calculate the best time to add hash power.
When we calculate the probability the variables used are Miner Hashrate, Network Hashrate, Block Time, and Mining Time.
There is no way to say "if I add more power now, I'll solve" because the block time is 14s. Every time you add hash power your probability per block increases but in general probability is a spread as nothing is ever 100%

**Reply by XIII39** [reply]
I put together a quick Excel sheet with Poisson and Binomial Distribution calculations.
Just enter your hashrate, the current network hashrate and days spent mining.
Also update the current price of ETH if you want to see the $$$

Just remember, mining POW is all about time. Consistency of block wins requires time, nothing more.
Mining with 100GH/s is good for 30 days but better with 60

https://docs.google.com/spreadsheets/d/e/2PACX-1vQNgPgjbKtMBUJfOAwLtSEPxRn1bn-xMmriM34VyBe_gj2IpnFzzXiqhSqzroB6nb9IE97F8KbCQN9o/pub?output=xlsx

You can use this for (almost) any POW mining where you know;
1. Your Hashrate
2. The Network Hashrate
3. Block Time
4. Your Mining Time

**Reply by alexanderkiehl** [reply]
You rock it @XIII39 Thanks
